### PR TITLE
Rename ApplicationModel to ApplicationProjectModel, add ApplicationIn…

### PR DESCRIPTION
**Pull request by OndraZizka**
…putPathModel, and mark the right ProjectModels and FileModels with those types.

### CODE REVIEW

File: graph/api/src/main/java/org/jboss/windup/graph/model/ApplicationArchiveModel.java [hunk 1]
```diff
@@ -3,7 +3,7 @@
 import com.tinkerpop.frames.modules.typedgraph.TypeValue;
 
 @TypeValue(ApplicationArchiveModel.TYPE)
-public interface ApplicationArchiveModel extends ApplicationModel, ArchiveModel
+public interface ApplicationArchiveModel extends ArchiveModel
```
Comment (jsight) [inline graph/api/src/main/java/org/jboss/windup/graph/model/ApplicationArchiveModel.java]: Why would this no longer extend ApplicationModel?

Comment (jsight) [reply graph/api/src/main/java/org/jboss/windup/graph/model/ApplicationArchiveModel.java]: Actually, I see why, but this all seems a bit strange now. It seems to me that ApplicationModel is a marker interface to indicate that something is an application.

Previously, (filemodel instanceof ApplicationModel) == true for cases where the file was an application.

After this change, that relationship is broken as you have redefined the meaning of "ApplicationModel".

File: graph/api/src/main/java/org/jboss/windup/graph/model/ApplicationProjectModel.java [hunk 2]
```diff
@@ -3,10 +3,10 @@
 import com.tinkerpop.frames.Property;
 import com.tinkerpop.frames.modules.typedgraph.TypeValue;
 
-@TypeValue(ApplicationModel.TYPE)
-public interface ApplicationModel extends WindupVertexFrame
+@TypeValue(ApplicationProjectModel.TYPE)
+public interface ApplicationProjectModel extends ProjectModel
```
Comment (jsight) [inline graph/api/src/main/java/org/jboss/windup/graph/model/ApplicationProjectModel.java]: This isn't just changing a name, this is changing the meaning, and it seems to me that the new meaning is less useful.

Why not just add the ApplicationModel marker interface to a ProjectModel vertex if you want to be able to differentiate them by type?

File: rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/scan/provider/DiscoverFilesAndTypesRuleProvider.java [hunk 3]
```diff
@@ -32,9 +36,22 @@ public Configuration getConfiguration(RuleLoaderContext ruleLoaderContext)
     {
         return ConfigurationBuilder.begin()
 
+        // Mark all input paths as ApplicationProjectModel.
```
Comment (jsight) [inline rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/scan/provider/DiscoverFilesAndTypesRuleProvider.java]: This comment is actually incorrect.

File: graph/api/src/main/java/org/jboss/windup/graph/model/ApplicationArchiveModel.java (hunk 1 above)
Comment (OndraZizka) [reply graph/api/src/main/java/org/jboss/windup/graph/model/ApplicationArchiveModel.java]: Yes, it's an marker interface. But the name is ambiguous, so I am creating ApplicationProjectModel, which marks a ProjectModel that is an app, and a ApplicationInputPathModel, which marks what is now FileModel instanceof ApplicationModel.

Comment (OndraZizka) [reply graph/api/src/main/java/org/jboss/windup/graph/model/ApplicationArchiveModel.java]: Note that a directory may also be an application, while not being an archive.

File: graph/api/src/main/java/org/jboss/windup/graph/model/ApplicationProjectModel.java (hunk 2 above)
Comment (OndraZizka) [reply graph/api/src/main/java/org/jboss/windup/graph/model/ApplicationProjectModel.java]: See above. I want to be able to ask for an app input paths or app projects, separatedly. Just marking them would not allow that. Plus, using it as general marker interface closes the possibility to have Project or File specific methods, relations and properties.

File: rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/scan/provider/DiscoverFilesAndTypesRuleProvider.java (hunk 3 above)
Comment (OndraZizka) [reply rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/scan/provider/DiscoverFilesAndTypesRuleProvider.java]: Right, this was changed by IDEA refactoring... fixed

File: graph/api/src/main/java/org/jboss/windup/graph/model/ApplicationProjectModel.java (hunk 2 above)
Comment (jsight) [reply graph/api/src/main/java/org/jboss/windup/graph/model/ApplicationProjectModel.java]: With the marker interface, you could have done "new GraphService<ApplicationModel>(event, ApplicationModel.class).findAll()". Now, that is impossible, so I don't see how getting rid of the ApplicationModel marker has helped you.

**Comment by jsight** [comment]
Instead of this, I would suggest the following hierarchy:

  - ApplicationFileModel
     - ApplicationArchiveModel
- ProjectModel
  - ApplicationProjectModel

In this way, ApplicationModel can still be a marker interface for all Application File objects and ProjectModel has a separate hierarchy.

The existing ApplicationModel would be renamed to ApplicationFileModel as that is really what it is. From what I can tell, the "ambiguity" that you are imaginging is the absence of an ApplicationProjectModel concept.

File: graph/api/src/main/java/org/jboss/windup/graph/model/ApplicationProjectModel.java (hunk 2 above)
Comment (OndraZizka) [reply graph/api/src/main/java/org/jboss/windup/graph/model/ApplicationProjectModel.java]: What is that good for? It is never used anywhere.

**Comment by OndraZizka** [comment]
Could be, perhaps. Isn't `ApplicationInputPath` a better matching name? It can be a directory, too. We don't even know if it is an existing file. We only know it's a path, initially.

**Comment by jsight** [comment]
@OndraZizka - Maybe ApplicationInputPathModel is ok, but is that always an accurate name? If the user gives us a directory as the input path and does not specify "exploded" then we will crawl the input directory and create applications from all files found there. Then we will have many application paths, but still only one InputPath. Wouldn't input path be misleading in that case? I agree that differentiating ApplicationArchive from ApplicationInputPath is probably unnecessary.

I was horribly unclear in my last sentence above. ApplicationFileModel (or ApplicationInputPathModel) would take the place of the existing ApplicationModel.

**Comment by OndraZizka** [comment]
Hmm good point. I'll think about it :)

**Pull request closed without merging**